mcs/class/System/System.Net.NetworkInformation/IPAddressCollection.cs

```diff
@@ -84,6 +84,25 @@ public static Win32IPAddressCollection FromDnsServer (IntPtr ptr)
 			return c;
 		}
 
+		public static Win32IPAddressCollection FromSocketAddress (Win32_SOCKET_ADDRESS addr)
+		{
+			Win32IPAddressCollection c = new Win32IPAddressCollection ();
+			if (addr.Sockaddr != IntPtr.Zero)
+				c.InternalAdd (addr.GetIPAddress ());
+			return c;
+		}
+
+		public static Win32IPAddressCollection FromWinsServer (IntPtr ptr)
+		{
+			Win32IPAddressCollection c = new Win32IPAddressCollection ();
+			Win32_IP_ADAPTER_WINS_SERVER_ADDRESS a;
+			for (IntPtr p = ptr; p != IntPtr.Zero; p = a.Next) {
+				a = (Win32_IP_ADAPTER_WINS_SERVER_ADDRESS) Marshal.PtrToStructure (p, typeof (Win32_IP_ADAPTER_WINS_SERVER_ADDRESS));
```

Comment:
Maybe do some sanity check that we at least have the expected struct length (included as a field in the unmanaged struct)?

Reply:
Did you want something like this?
```
		public static Win32IPAddressCollection FromWinsServer (IntPtr ptr)
		{
			Win32IPAddressCollection c = new Win32IPAddressCollection ();
			Win32_IP_ADAPTER_WINS_SERVER_ADDRESS a;
			for (IntPtr p = ptr; p != IntPtr.Zero; p = a.Next) {
				a = (Win32_IP_ADAPTER_WINS_SERVER_ADDRESS) Marshal.PtrToStructure (p, typeof (Win32_IP_ADAPTER_WINS_SERVER_ADDRESS));
				if(a.LengthFlags.Length != Marshal.SizeOf(typeof(Win32_IP_ADAPTER_WINS_SERVER_ADDRESS))) 
						return new Win32IPAddressCollection ();

				c.InternalAdd (a.Address.GetIPAddress ());
			}
			return c;
		}
```
Then if any of our list of Win32_IP_ADAPTER_WINS_SERVER_ADDRESS are junk than we'll return an empty address collection.

Reply:
So the check should be that we don't get a a.LengthFlags.Length that is smaller than what we expect (meaning that we are on an earlier unsupported version of Windows). I did a check in corefx and it doesn't look like they do this additional checks though so maybe it is overkill. I'm fine leaving it as is.